kv/kvnemesis: TestKVNemesisMultiNode failed

kv/kvnemesis.TestKVNemesisMultiNode [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/14169943?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/14169943?buildTab=artifacts#/) on release-23.1 @ [5d02fdf2851038279f7544e6271a08e1036a2966](https://github.com/cockroachdb/cockroach/commits/5d02fdf2851038279f7544e6271a08e1036a2966):


```
=== RUN   TestKVNemesisMultiNode
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisMultiNode681426315
    test_log_scope.go:79: use -show-logs to present logs inline
    kvnemesis_test.go:180: seed: 8246618713428840787
    kvnemesis_test.go:124: kvnemesis logging to /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis2506896689
    kvnemesis.go:165: error applying x.ScanForUpdate(ctx, tk(7627554184111334918), tk(11297944047384909700), 0) // WriteTooOldError: write for key /Table/100/"83f343408bfd2cf2" at timestamp 1709031554.741791117,0 too old; wrote at 1709031554.784361740,1: {<nil> 0 {0xc002fbb670} <nil> 1709031554.919155299,0}
    kvnemesis.go:185: failures(verbose): /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis2506896689/failures
        repro steps: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis2506896689/repro.go
        rangefeed KVs: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis2506896689/kvs-rangefeed.txt
        scan KVs: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis2506896689/kvs-scan.txt
    kvnemesis_test.go:207: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:207
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:160
        	Error:      	Should be zero, but was 1
        	Test:       	TestKVNemesisMultiNode
        	Messages:   	kvnemesis detected failures
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisMultiNode681426315
--- FAIL: TestKVNemesisMultiNode (19.24s)
```

Parameters:
 - <code>TAGS=bazel,gss</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #118471 kv/kvnemesis: TestKVNemesisMultiNode failed [C-test-failure O-robot P-2 T-kv X-duplicate branch-release-23.2]
- #118017 kv/kvnemesis: TestKVNemesisMultiNode failed [A-testing C-bug C-test-failure O-robot P-2 T-kv branch-master]
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisMultiNode.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-36279

--- a/pkg/storage/pebble_mvcc_scanner.go
+++ b/pkg/storage/pebble_mvcc_scanner.go
@@ -551,7 +551,13 @@ func (p *pebbleMVCCScanner) init(
 	// because the local uncertainty limit cannot be applied to values with
 	// future-time timestamps with earlier local timestamps. We are only able
 	// to skip uncertainty checks if p.ts >= global_uncertainty_limit.
-	p.checkUncertainty = p.ts.Less(p.uncertainty.GlobalLimit)
+	//
+	// We disable checkUncertainty when the scanner is configured with failOnMoreRecent.
+	// This avoids cases in which a scan would have failed with a WriteTooOldError
+	// but instead gets an unexpected ReadWithinUncertaintyIntervalError
+	// See:
+	// https://github.com/cockroachdb/cockroach/issues/119681
+	p.checkUncertainty = p.ts.Less(p.uncertainty.GlobalLimit) && !p.failOnMoreRecent
 }
 
 // get seeks to the start key exactly once and adds one KV to the result set.

--- a/pkg/storage/testdata/mvcc_histories/range_tombstone_scans
+++ b/pkg/storage/testdata/mvcc_histories/range_tombstone_scans
@@ -477,6 +477,18 @@ scan k=b end=d ts=3 globalUncertaintyLimit=4
 scan: "b"-"d" -> <no data>
 error: (*kvpb.ReadWithinUncertaintyIntervalError:) ReadWithinUncertaintyIntervalError: read at time 3.000000000,0 encountered previous write with future timestamp 4.000000000,0 within uncertainty interval `t <= (local=0,0, global=0,0)`; observed timestamps: []
 
+run error
+scan k=b end=d ts=3 globalUncertaintyLimit=4 failOnMoreRecent
+----
+scan: "b"-"d" -> <no data>
+error: (*kvpb.ReadWithinUncertaintyIntervalError:) ReadWithinUncertaintyIntervalError: read at time 3.000000000,0 encountered previous write with future timestamp 4.000000000,0 within uncertainty interval `t <= (local=0,0, global=0,0)`; observed timestamps: []
+
+run error
+get k=c ts=3 globalUncertaintyLimit=4 failOnMoreRecent
+----
+get: "c" -> <no data>
+error: (*kvpb.ReadWithinUncertaintyIntervalError:) ReadWithinUncertaintyIntervalError: read at time 3.000000000,0 encountered previous write with future timestamp 4.000000000,0 within uncertainty interval `t <= (local=0,0, global=0,0)`; observed timestamps: []
+
 run ok
 scan k=b end=d ts=4 globalUncertaintyLimit=5
 ----

--- a/pkg/storage/testdata/mvcc_histories/range_tombstone_scans
+++ b/pkg/storage/testdata/mvcc_histories/range_tombstone_scans
@@ -481,13 +481,13 @@ run error
 scan k=b end=d ts=3 globalUncertaintyLimit=4 failOnMoreRecent
 ----
 scan: "b"-"d" -> <no data>
-error: (*kvpb.ReadWithinUncertaintyIntervalError:) ReadWithinUncertaintyIntervalError: read at time 3.000000000,0 encountered previous write with future timestamp 4.000000000,0 within uncertainty interval `t <= (local=0,0, global=0,0)`; observed timestamps: []
+error: (*kvpb.WriteTooOldError:) WriteTooOldError: write for key "c" at timestamp 3.000000000,0 too old; must write at or above 4.000000000,1
 
 run error
 get k=c ts=3 globalUncertaintyLimit=4 failOnMoreRecent
 ----
 get: "c" -> <no data>
-error: (*kvpb.ReadWithinUncertaintyIntervalError:) ReadWithinUncertaintyIntervalError: read at time 3.000000000,0 encountered previous write with future timestamp 4.000000000,0 within uncertainty interval `t <= (local=0,0, global=0,0)`; observed timestamps: []
+error: (*kvpb.WriteTooOldError:) WriteTooOldError: write for key "c" at timestamp 3.000000000,0 too old; must write at or above 4.000000000,1
 
 run ok
 scan k=b end=d ts=4 globalUncertaintyLimit=5

